src/site/content/en/blog/device-pixel-content-box/index.md

```diff
@@ -0,0 +1,135 @@
+
+```html
+<style>
+  /* … styles that affect the canvas' size … */
+</style>
+<canvas id="myCanvas"></canvas>
+<script>
+  const cvs = document.querySelector('#myCanvas');
+  // Get the canvas' size in CSS pixels
+  const rectangle = cvs.getBoundingClientRect();
+  // Convert it to real pixels. Ish.
+  cvs.width = rectangle.width * devicePixelRatio;
+  cvs.height = rectangle.height * devicePixelRatio;
+  // Start drawing…
+</script>
+```
+
+The astute reader might be wondering what happens when dPR is not an integer value. That is a good question and exactly where the crux of this entire problem lies. In addition, if you specify an element's position or size using percentages, `vh`, or other indirect values, it is possible that they will resolve to fractional CSS pixel values. An element with `margin-left: 33%` can end up with a rectangle like this:
+
+<figure class="w-figure">
+  <img src="fractional-pixels.png" alt="" class="w-screenshot" width="600">
+  <figcaption class="w-figcaption">Dev Tools showing fractional pixel values as a result of a <code>getBoundingClientRect()</code> call.</figcaption>
+</figure>
+
+CSS pixels are purely virtual, so having fractions of a pixel is okay in theory, but how does the browser figure out the mapping to physical pixels? Because fractional _physical_ pixels are not a thing.
+
+## Pixel Snapping
+
+The part of the unit conversion process that takes care of aligning elements with physical pixels is called "pixel snapping", and it does what it says on the tin: It snaps fractional pixel values to integer, physical pixel values. How exactly this happens is different from browser to browser. If we have an element with a width of `791.984px` on a display where dPR is 1, one browser might render the element at `792px` physical pixels, while another browser might render it at `791px`. That's just a single pixel off, but a single pixel can be detrimental to renderings that need to be pixel-perfect. This can lead to blurriness or even more visible artifacts like the [Moiré effect].
+
+<figure class="w-figure">
+  <img src="side-by-side.png" alt="" class="w-screenshot" width="600">
+  <figcaption class="w-figcaption">The top image is a raster of differently colored pixels. The bottom image is the same as above, but the width and height have been reduced by one pixel using bilinear scaling. The emerging pattern is called the Moiré effect.<br>(You might have to open this image in a new tab to see it without any scaling applied to it.)</figcaption>
+</figure>
+
+## devicePixelContentBox
+
+At this point, `devicePixelContentBox` should make a lot more sense to you. It gives you an element's content box in device pixel (i.e. physical pixel) units. It's part of `ResizeObserver`. While [ResizeObserver is now supported in all major browsers][ro support] since Safari 13.1, the `devicePixelContentBox` property is only in Chrome 84+ for now.
```

Comment:
```suggestion
`devicePixelContentBox` gives you an element's content box in device pixel (i.e. physical pixel) units. It's part of `ResizeObserver`. While [ResizeObserver is now supported in all major browsers][ro support] since Safari 13.1, the `devicePixelContentBox` property is only in Chrome 84+ for now.
```